sql: Add support for SHOW CHANGEFEED JOBS

spiffyy99 opened the pull request:
Previously, SHOW JOBS did not contain enough changefeed information
for users.
SHOW CHANGEFEED JOBS only shows changefeed jobs as well as the full
table name the changefeed monitors, the changefeed sink URI and the
format as computed columns.

Resolves #64933

Release note (sql change): Added new SHOW CHANGEFEED JOBS command with
additional information about changefeeds for improved user visibility

Specifically, `SHOW CHANGEFEED JOBS` adds columns `sink_uri`, 
`full_table_names`, and `format` as well as `high_water_timestamp` from 
`crdb_internal.jobs`.
![image](https://user-images.githubusercontent.com/24757058/117721773-42ea2880-b1ae-11eb-8a05-e150ff2a2fef.png)

Some of the metrics we wanted to include in here can't be added (i.e. buffer size, # of errors) since they aren't shown on an individual changefeed basis, can create another ticket for that once it is ready.

stevendanna reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64956)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @spiffyyeng)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/64956#-M_QG3WE8GYcolIvI8-K:-M_QG3WE8GYcolIvI8-L:b-4yilkz) (no related file):*
Thanks for working on this! I think we and our customers are going to get a lot of use out of it going forward.

I left a few minor comments that I hope aren't too annoying.  Overall though it looks pretty reasonable.  I didn't review the actual query being used too closely since we are going to iterate on what information gets included there over time.

---
___
*[pkg/jobs/jobs_test.go, line 1732 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64956#-M_Q7Ig-CnoHF_zF8V4k:-M_Q7Ig-CnoHF_zF8V4l:bafy8tp) ([raw file](https://github.com/cockroachdb/cockroach/blob/09ed83a2ef082ea738b35b6fe3482ec7a0aab0d3/pkg/jobs/jobs_test.go#L1732)):*
> ```Go
> 	rowResults := sqlDB.Query(t, `SELECT job_id, job_type, sink_uri, full_table_names, format FROM [SHOW CHANGEFEED JOBS]`)
> 
> 	rowResults.Next()
> ```

The sql.Rows interface requires that if Next() returns false that Err() is checked.  In this case, it likely doesn't matter much, as the test will still fail at the Scan attempt, but it might produce slightly easier-to-debug failures if we get the error. Perhaps something like (warning, completely untested):

```go
	if !rowResults.Next() {
		t.Fatalf("row expected but not found: %v", rowResult.Err())
	}
	err := rowResults.Scan(&out.id, &out.typ, &out.SinkURI, &out.FulLTableNames, &out.format)
	if err != nil {
		t.Fatal(err)
	}
```
___
*[pkg/jobs/jobs_test.go, line 1737 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64956#-M_Q96GT8RKOsMi6o7MR:-M_Q96GT8RKOsMi6o7MS:b-lmubq2) ([raw file](https://github.com/cockroachdb/cockroach/blob/09ed83a2ef082ea738b35b6fe3482ec7a0aab0d3/pkg/jobs/jobs_test.go#L1737)):*
> ```Go
> 		t.Fatal(err)
> 	}
> 	if out.id != 1 ||
> ```

We are asserting the order of the results here, but I don't see an ORDER BY in the query that drives SHOW CHANGEFEED JOBS or in the SELECT above.  It almost surely doesn't matter given the table and that this is a single-node test, but might be nice to put an order by on one of them
___
*[pkg/sql/delegate/show_jobs.go, line 41 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64956#-M_QFkjM6oVqTeQXICTl:-M_QFkjN9yg7GtGT_Va8:b-72wv8o) ([raw file](https://github.com/cockroachdb/cockroach/blob/09ed83a2ef082ea738b35b6fe3482ec7a0aab0d3/pkg/sql/delegate/show_jobs.go#L41)):*
> ```Go
>                SELECT job_id, job_type, description, statement, user_name, status,
>                running_status, created, started, finished, modified,
>                fraction_completed, high_water_timestamp, error, coordinator_id, 
> ```

Doesn't need to change in this PR, but I think we found that coordinator_id is always NULL.  Since this is a new command where we can change the output, perhaps we can remove it if we expect it to stay NULL?  (Perhaps also fraction_completed and statement?)
___
*[pkg/sql/delegate/show_jobs.go, line 45 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64956#-M_QEItQ28kbVrlcad57:-M_QEItR5LOMXm0mvHtd:b-nq90up) ([raw file](https://github.com/cockroachdb/cockroach/blob/09ed83a2ef082ea738b35b6fe3482ec7a0aab0d3/pkg/sql/delegate/show_jobs.go#L45)):*
> ```Go
>                ARRAY(SELECT concat(database_name, '.', schema_name, '.', name) from crdb_internal.tables WHERE table_id = any(descriptor_ids)) AS full_table_names,
>                btrim(jsonb_pretty(json_extract_path(payload_json, 'changefeed', 'opts', 'format')), '"') AS format
> 				FROM crdb_internal.jobs INNER JOIN payload ON id = job_id`
> ```

nit: I think future us might appreciate a little bit more formatting here.  Perhaps something like has been done in show_table.go?


<!-- Sent from Reviewable.io -->

miretskiy reviewed (changes requested):
> Reviewed 1 of 6 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64956)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @spiffyyeng and @stevendanna)

___
*[pkg/sql/delegate/show_jobs.go, line 45 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64956#-M_QEItQ28kbVrlcad57:-M_R0kGX1Yd7OCV1YMeN:b-2c64kt) ([raw file](https://github.com/cockroachdb/cockroach/blob/09ed83a2ef082ea738b35b6fe3482ec7a0aab0d3/pkg/sql/delegate/show_jobs.go#L45)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

nit: I think future us might appreciate a little bit more formatting here.  Perhaps something like has been done in show_table.go?
</blockquote></details>

I'll +1 this.  
___
*[pkg/sql/delegate/show_jobs.go, line 74 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64956#-M_R1mxr0NTkzrZ35oRp:-M_R1mxr0NTkzrZ35oRq:b-11h7bj) ([raw file](https://github.com/cockroachdb/cockroach/blob/09ed83a2ef082ea738b35b6fe3482ec7a0aab0d3/pkg/sql/delegate/show_jobs.go#L74)):*
> ```Go
> 	var sqlStmt string
> 	if n.WhichJobs == tree.ChangefeedJobs {
> 		sqlStmt = fmt.Sprintf("%s %s %s", selectClauseChangefeed, whereClause, orderbyClause)
> ```

You probably need to update parse tests: `pkg/sql/parser/testdata/show`
___
*[pkg/sql/sem/tree/show.go, line 255 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64956#-M_R26qP1hz6H1hOJQQD:-M_R26qP1hz6H1hOJQQE:b-562a2k) ([raw file](https://github.com/cockroachdb/cockroach/blob/09ed83a2ef082ea738b35b6fe3482ec7a0aab0d3/pkg/sql/sem/tree/show.go#L255)):*
> ```Go
> 
> const (
> ```

dro extra \n please
___
*[pkg/sql/sem/tree/show.go, line 302 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64956#-M_R2JpA77HIePcpJLmp:-M_R2JpA77HIePcpJLmq:bdckgep) ([raw file](https://github.com/cockroachdb/cockroach/blob/09ed83a2ef082ea738b35b6fe3482ec7a0aab0d3/pkg/sql/sem/tree/show.go#L302)):*
> ```Go
> 	ctx.WriteString("SHOW ")
> 	if node.WhichJobs != NonAutomaticJobs {
> 		ctx.FormatNode(&node.WhichJobs)
> ```

you can probably call format node without above "if" check (you have default clause in job type format)


<!-- Sent from Reviewable.io -->

spiffyy99 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64956)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @spiffyyeng, and @stevendanna)

___
*[pkg/jobs/jobs_test.go, line 1732 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64956#-M_Q7Ig-CnoHF_zF8V4k:-M_RPiTT14EdvA_2S9XU:b-b2wvqe) ([raw file](https://github.com/cockroachdb/cockroach/blob/09ed83a2ef082ea738b35b6fe3482ec7a0aab0d3/pkg/jobs/jobs_test.go#L1732)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

The sql.Rows interface requires that if Next() returns false that Err() is checked.  In this case, it likely doesn't matter much, as the test will still fail at the Scan attempt, but it might produce slightly easier-to-debug failures if we get the error. Perhaps something like (warning, completely untested):

```go
	if !rowResults.Next() {
		t.Fatalf("row expected but not found: %v", rowResult.Err())
	}
	err := rowResults.Scan(&out.id, &out.typ, &out.SinkURI, &out.FulLTableNames, &out.format)
	if err != nil {
		t.Fatal(err)
	}
```

</blockquote></details>

good catch. looks like Err() actually returns nil if there's no error (could just be no rows left), so accounted for this as well.
___
*[pkg/jobs/jobs_test.go, line 1737 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64956#-M_Q96GT8RKOsMi6o7MR:-M_RPzUT0g4G6QI78O7U:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/09ed83a2ef082ea738b35b6fe3482ec7a0aab0d3/pkg/jobs/jobs_test.go#L1737)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

We are asserting the order of the results here, but I don't see an ORDER BY in the query that drives SHOW CHANGEFEED JOBS or in the SELECT above.  It almost surely doesn't matter given the table and that this is a single-node test, but might be nice to put an order by on one of them
</blockquote></details>

Done.
___
*[pkg/sql/delegate/show_jobs.go, line 41 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64956#-M_QFkjM6oVqTeQXICTl:-M_RQ2ku7cAwQvZw5lhj:b-fmh20e) ([raw file](https://github.com/cockroachdb/cockroach/blob/09ed83a2ef082ea738b35b6fe3482ec7a0aab0d3/pkg/sql/delegate/show_jobs.go#L41)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Doesn't need to change in this PR, but I think we found that coordinator_id is always NULL.  Since this is a new command where we can change the output, perhaps we can remove it if we expect it to stay NULL?  (Perhaps also fraction_completed and statement?)
</blockquote></details>

i'll create a ticket
___
*[pkg/sql/sem/tree/show.go, line 255 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64956#-M_R26qP1hz6H1hOJQQD:-M_RQPMr9IYpLMr_zPtV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/09ed83a2ef082ea738b35b6fe3482ec7a0aab0d3/pkg/sql/sem/tree/show.go#L255)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

dro extra \n please
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/show.go, line 302 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64956#-M_R2JpA77HIePcpJLmp:-M_RQ9_94XUKVLkOzdxB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/09ed83a2ef082ea738b35b6fe3482ec7a0aab0d3/pkg/sql/sem/tree/show.go#L302)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

you can probably call format node without above "if" check (you have default clause in job type format)
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

miretskiy reviewed (changes requested):
> Reviewed 2 of 6 files at r1, 2 of 5 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64956)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @spiffyyeng and @stevendanna)

___
*[pkg/ccl/changefeedccl/changefeedbase/options.go, line 95 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/64956#-M_RYnjJBbpn6Cxg6OHj:-M_RYnjJBbpn6Cxg6OHk:b-svsjdy) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba75059382ae3cf16ae6413a3f295bf1982578b1/pkg/ccl/changefeedccl/changefeedbase/options.go#L95)):*
> ```Go
> 	SinkParamFileSize           = `file_size`
> 	SinkParamRegistryClientCert = `registry_client_cert`
> 	SinkParamRegistryClientKey  = `registry_client_cert`
> ```

why did we change these options?
___
*[pkg/jobs/jobs_test.go, line 1650 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/64956#-M_RYvdM0paJZXnkYcgb:-M_RYvdM0paJZXnkYcgc:b-k9itzy) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba75059382ae3cf16ae6413a3f295bf1982578b1/pkg/jobs/jobs_test.go#L1650)):*
> ```Go
> }
> 
> func TestShowChangefeedJobs(t *testing.T) {
> ```

I honestly don't think this test belongs here.  I think it belongs in the changefeedccl package.
___
*[pkg/jobs/jobs_test.go, line 1724 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/64956#-M_RZBTEEE-8kw1XlFby:-M_RZBTEEE-8kw1XlFbz:bqti59t) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba75059382ae3cf16ae6413a3f295bf1982578b1/pkg/jobs/jobs_test.go#L1724)):*
> ```Go
> 		sqlDB.Exec(t,
> 			`INSERT INTO system.jobs (id, status, payload) VALUES ($1, $2, $3)`,
> 			in.id, in.status, inPayload,
> ```

I'm not too thrilled about inserting jobs like this.  The registry could adopt this job; and start running it for example.

I think it would be better if we executed "CREATE CHANGEFEED" statement instead.
There are few ways we can prevent those "fake" changefeeds from being executed.  But the one that I think makes most sense is to muck with TestingResumerCreationKnobs -- see example in `import_stmt_test` (sets up resumer for import statement).
Basically, you could write a "fake changefeed" resumer, which probably blocks until you cancel context.
After you created changefeed, you can examine it.
___
*[pkg/jobs/jobs_test.go, line 1753 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/64956#-M_R_GEz-xQdvhHnPPIa:-M_R_GEz-xQdvhHnPPIb:bn7fqzi) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba75059382ae3cf16ae6413a3f295bf1982578b1/pkg/jobs/jobs_test.go#L1753)):*
> ```Go
> 			"but found id:%d, type:%s, sinkUri:%s, fullTableNames:%s and format:%s",
> 			1, "CHANGEFEED", "kafka://localhost:9092", "{defaultdb.public.foo}", "json",
> 			out.id, out.typ, out.SinkURI, string(out.FulLTableNames), out.format)
> ```

I like using `require.Equal` -- this prints expected/found automatically, and you can add extra information too
`require.Equal(t, "json", out.format, "extra error information: %s ....", ...)`
___
*[pkg/jobs/jobs_test.go, line 1777 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/64956#-M_R_8Bg2300xqV37azv:-M_R_8Bg2300xqV37azw:btitmp2) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba75059382ae3cf16ae6413a3f295bf1982578b1/pkg/jobs/jobs_test.go#L1777)):*
> ```Go
> 			2, "CHANGEFEED", "experimental-s3://fake-bucket-name/fake/path", "{defaultdb.public.foo,defaultdb.public.bar}", "avro",
> 			out.id, out.typ, out.SinkURI, string(out.FulLTableNames), out.format)
> 	}
> ```

We are also missing tests when you try to `show changefeed job 123` when the job is not a changefeed job.
We should also test how the statement behaves when the job is paused, cancelled, etc.
___
*[pkg/sql/delegate/show_jobs.go, line 72 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/64956#-M_RX5X98HR4QSJUttX4:-M_RX5X98HR4QSJUttX5:bpn9wvg) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba75059382ae3cf16ae6413a3f295bf1982578b1/pkg/sql/delegate/show_jobs.go#L72)):*
> ```Go
>     ), 
>     '"'
>   ) AS sink_uri, 
> ```

This is probably not okay.  We need to make sure we don't show sensitive
information here -- and URIs can contain e.g. access keys.

For example, see `func (m ScheduledBackupExecutionArgs) MarshalJSONPB(x *jsonpb.Marshaler) ([]byte, error) {`
(in `created_scheduled_backup`) -- we had to write custom marshaller so that we won't display those secrets.
___
*[pkg/sql/delegate/show_jobs.go, line 97 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/64956#-M_RYM3E7uO_JJ9GE29y:-M_RYM3E7uO_JJ9GE29z:bl4wi81) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba75059382ae3cf16ae6413a3f295bf1982578b1/pkg/sql/delegate/show_jobs.go#L97)):*
> ```Go
>   INNER JOIN payload ON id = job_id`
> 	)
> ```

This query is LARGE.  And it's likely that it will grow larger still.
What's also likely, that it might no longer be feasible (or convenient) to use a 1 shot query to display
everything we may want to display.  As such, I think it might make sense to move these changes into e.g. `show_changefeed.go` file.  I believe you should be able to introduce `tree.ShowChangefeed` node (like ShowSchedules one).  


<!-- Sent from Reviewable.io -->

rytaft commented:
Looks like you need to rebase -- you've pulled in a lot of unrelated commits.





<!-- Sent from Reviewable.io -->

spiffyyeng commented:
fixed now, not sure what happened there 😂

spiffyy99 reviewed (commented):
> Dismissed @miretskiy from 3 discussions.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64956)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @spiffyyeng, and @stevendanna)

___
*[pkg/jobs/jobs_test.go, line 1650 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/64956#-M_RYvdM0paJZXnkYcgb:-Ma95imN7Au-dSseWMwK:b-x73075) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba75059382ae3cf16ae6413a3f295bf1982578b1/pkg/jobs/jobs_test.go#L1650)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I honestly don't think this test belongs here.  I think it belongs in the changefeedccl package.
</blockquote></details>

moved to changefeedccl
___
*[pkg/jobs/jobs_test.go, line 1724 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/64956#-M_RZBTEEE-8kw1XlFby:-Ma95uz61FW7XGioZ0Ip:b-n9tkhd) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba75059382ae3cf16ae6413a3f295bf1982578b1/pkg/jobs/jobs_test.go#L1724)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I'm not too thrilled about inserting jobs like this.  The registry could adopt this job; and start running it for example.

I think it would be better if we executed "CREATE CHANGEFEED" statement instead.
There are few ways we can prevent those "fake" changefeeds from being executed.  But the one that I think makes most sense is to muck with TestingResumerCreationKnobs -- see example in `import_stmt_test` (sets up resumer for import statement).
Basically, you could write a "fake changefeed" resumer, which probably blocks until you cancel context.
After you created changefeed, you can examine it.

</blockquote></details>

changed
___
*[pkg/sql/delegate/show_jobs.go, line 74 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64956#-M_R1mxr0NTkzrZ35oRp:-M_RRvr89x-zh8Q8_YxZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/09ed83a2ef082ea738b35b6fe3482ec7a0aab0d3/pkg/sql/delegate/show_jobs.go#L74)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

You probably need to update parse tests: `pkg/sql/parser/testdata/show`
</blockquote></details>

Done.
___
*[pkg/sql/delegate/show_jobs.go, line 72 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/64956#-M_RX5X98HR4QSJUttX4:-Ma96B1W8XINFXgSlvTj:b-jgjwcz) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba75059382ae3cf16ae6413a3f295bf1982578b1/pkg/sql/delegate/show_jobs.go#L72)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

This is probably not okay.  We need to make sure we don't show sensitive
information here -- and URIs can contain e.g. access keys.

For example, see `func (m ScheduledBackupExecutionArgs) MarshalJSONPB(x *jsonpb.Marshaler) ([]byte, error) {`
(in `created_scheduled_backup`) -- we had to write custom marshaller so that we won't display those secrets.
</blockquote></details>

added this method for ChangfeedDetails, sinkuri is now redacted
___
*[pkg/sql/delegate/show_jobs.go, line 97 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/64956#-M_RYM3E7uO_JJ9GE29y:-Ma96Jo8CiziDwGLidfs:b-p6ajuu) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba75059382ae3cf16ae6413a3f295bf1982578b1/pkg/sql/delegate/show_jobs.go#L97)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

This query is LARGE.  And it's likely that it will grow larger still.
What's also likely, that it might no longer be feasible (or convenient) to use a 1 shot query to display
everything we may want to display.  As such, I think it might make sense to move these changes into e.g. `show_changefeed.go` file.  I believe you should be able to introduce `tree.ShowChangefeed` node (like ShowSchedules one).  
</blockquote></details>

moved and simplified the query a bit


<!-- Sent from Reviewable.io -->

miretskiy reviewed (changes requested):
> Reviewed 13 of 15 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64956)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @spiffyyeng, and @stevendanna)

___
*[pkg/ccl/changefeedccl/show_changefeed_jobs_test.go, line 28 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/64956#-Ma9TnX6851aKj_uL-IL:-Ma9TnX6851aKj_uL-IM:b-cee02w) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ce2439c0444adeff194f1bf8dacbbb8e6fe428f/pkg/ccl/changefeedccl/show_changefeed_jobs_test.go#L28)):*
> ```Go
> 	"github.com/stretchr/testify/require"
> )
> ```

Overall -- good job on tests.  Few nits/improvements here.
___
*[pkg/ccl/changefeedccl/show_changefeed_jobs_test.go, line 48 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/64956#-Ma9SmkcC9_pVoLSE-4L:-Ma9SmkcC9_pVoLSE-4M:bbo5gv6) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ce2439c0444adeff194f1bf8dacbbb8e6fe428f/pkg/ccl/changefeedccl/show_changefeed_jobs_test.go#L48)):*
> ```Go
> }
> 
> func waitForJobOperation(
> ```

nit: i'd call it waitForJobStatus
___
*[pkg/ccl/changefeedccl/show_changefeed_jobs_test.go, line 115 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/64956#-Ma9SyOL5qOyEX1JWIhs:-Ma9SyOL5qOyEX1JWIht:bmu284h) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ce2439c0444adeff194f1bf8dacbbb8e6fe428f/pkg/ccl/changefeedccl/show_changefeed_jobs_test.go#L115)):*
> ```Go
> 	// Cannot use kafka for tests right now because of leaked goroutine issue
> 	query = `CREATE CHANGEFEED FOR TABLE foo, bar INTO 
> 		'experimental-s3://fake-bucket-name/fake/path?AWS_ACCESS_KEY_ID=123&AWS_SECRET_ACCESS_KEY=456'`
> ```

the query itself returns jobID -- you should read that instead of querying jobs table below.
___
*[pkg/ccl/changefeedccl/show_changefeed_jobs_test.go, line 204 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/64956#-Ma9TOYI3kl_nRtH5g53:-Ma9TOYI3kl_nRtH5g54:b-oykztr) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ce2439c0444adeff194f1bf8dacbbb8e6fe428f/pkg/ccl/changefeedccl/show_changefeed_jobs_test.go#L204)):*
> ```Go
> 	sqlDB.Exec(t, query, changefeedID)
> 
> 	waitForJobOperation(sqlDB, t, changefeedID, "paused")
> ```

you probably want to run `show changefeed` statements again after pausing/cancelling (and also you need to resume too).
You should verify that show changefeed job does something reasonable.  Otherwise, why bother pausing/resuming?  You're not 
testing jobs system here.
___
*[pkg/ccl/changefeedccl/show_changefeed_jobs_test.go, line 212 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/64956#-Ma9TCwN4L5uFOAseDsb:-Ma9TCwN4L5uFOAseDsc:bnmo2p0) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ce2439c0444adeff194f1bf8dacbbb8e6fe428f/pkg/ccl/changefeedccl/show_changefeed_jobs_test.go#L212)):*
> ```Go
> }
> 
> func TestShowChangefeedJobsNoResults(t *testing.T) {
> ```

I think you should also create a different type of job (say backup), but run `show changefeed job` on that.
Make sure you get sane results.
___
*[pkg/sql/delegate/show_changefeed_jobs.go, line 26 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/64956#-Ma9VsyH4Gn53Oa_B2nL:-Ma9VsyH4Gn53Oa_B2nM:b-vx02wi) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ce2439c0444adeff194f1bf8dacbbb8e6fe428f/pkg/sql/delegate/show_changefeed_jobs.go#L26)):*
> ```Go
> 	const (
> 		selectClause = `
> WITH payload AS (
> ```

I'm still 
___
*[pkg/sql/delegate/show_changefeed_jobs.go, line 38 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/64956#-Ma9B4VG8--adWOJfEoH:-Ma9B4VG8--adWOJfEoI:b-m9xi98) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ce2439c0444adeff194f1bf8dacbbb8e6fe428f/pkg/sql/delegate/show_changefeed_jobs.go#L38)):*
> ```Go
> SELECT 
>   job_id, 
>   job_type, 
> ```

probably not needed since you know it's a changefeed job?
___
*[pkg/sql/delegate/show_changefeed_jobs.go, line 49 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/64956#-Ma9B1dm7n00rn5opSO3:-Ma9B1dm7n00rn5opSO4:b-9iav1) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ce2439c0444adeff194f1bf8dacbbb8e6fe428f/pkg/sql/delegate/show_changefeed_jobs.go#L49)):*
> ```Go
>   high_water_timestamp, 
>   error, 
> 	replace(
> ```

nit: formatting a bit off?
___
*[pkg/sql/delegate/show_changefeed_jobs.go, line 52 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/64956#-Ma9AxhIAhnTqn8FECfT:-Ma9AxhIAhnTqn8FECfU:b-5vz59q) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ce2439c0444adeff194f1bf8dacbbb8e6fe428f/pkg/sql/delegate/show_changefeed_jobs.go#L52)):*
> ```Go
> 		changefeed_details->>'sink_uri',
> 		'\u0026', '&'
> 	) AS sink_uri, 
> ```

doesn't sink_uri need to be redacted? 
Ahh... I see.  You have changefeed_details produced as pb_to_json -- and that did the redaction.
Perhaps add a comment?
___
*[pkg/sql/delegate/show_changefeed_jobs.go, line 66 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/64956#-Ma9BOPk9Z0Ai1krtHCV:-Ma9BOPk9Z0Ai1krtHCW:bscqf45) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ce2439c0444adeff194f1bf8dacbbb8e6fe428f/pkg/sql/delegate/show_changefeed_jobs.go#L66)):*
> ```Go
> 	changefeed_details->'opts'->>'format' AS format 
> FROM 
>   crdb_internal.jobs 
> ```

I wonder why we need to query crdb_internal.jobs at all?
Actually, scratch that -- there is reason why we want to query that instead of system.jobs -- namely permission checking.
But that makes me a bit sad (cause I'm not a fan of crdb_internal.jobs)

Long time ago, I tried rewriting show jobs, and I was able to emulate permission checking directly in sql by doing something like:
```
+       const decodedJobs = `
+SELECT
+   id, status, created,
+   crdb_internal.pb_to_json('cockroach.sql.jobs.jobspb.Payload', payload) as payload,
+   crdb_internal.pb_to_json('cockroach.sql.jobs.jobspb.Progress', progress) as progress
+FROM system.jobs

+  permissions AS (
+    SELECT
+      crdb_internal.is_admin() as isAdmin,
+      crdb_internal.has_role_option('CONTROLJOB') as hasJobControl
+  ),
+  jobs AS (
+    SELECT *, %s
+    FROM (WITH decodedJobs AS (%s %s) SELECT * FROM decodedJobs), permissions
+    WHERE permissions.isAdmin OR permissions.hasJobControl OR payload->>'username' = current_user()
+  )

```

I don't know.  Let me think a bit more about this query, and whether we should just not use crdb_internal.jobs.
For now, let's address other comments, but we might have to come back to this particular query.
___
*[pkg/sql/sem/tree/stmt.go, line 1254 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/64956#-Ma9AFi58cgwZ_aSH9_z:-Ma9AFi58cgwZ_aSH9a-:b-qxt80x) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ce2439c0444adeff194f1bf8dacbbb8e6fe428f/pkg/sql/sem/tree/stmt.go#L1254)):*
> ```Go
> 
> // StatementTag returns a short string identifying the type of statement.
> func (*ShowChangefeedJobs) StatementTag() string { return "SHOW JOBS" }
> ```

`SHOW CHANGEFEED JOBS`?


<!-- Sent from Reviewable.io -->

spiffyy99 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64956)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @spiffyyeng, and @stevendanna)

___
*[pkg/sql/delegate/show_changefeed_jobs.go, line 66 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/64956#-Ma9BOPk9Z0Ai1krtHCV:-Ma9gMV37zeJp1jOSxd7:b-2uhanc) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ce2439c0444adeff194f1bf8dacbbb8e6fe428f/pkg/sql/delegate/show_changefeed_jobs.go#L66)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I wonder why we need to query crdb_internal.jobs at all?
Actually, scratch that -- there is reason why we want to query that instead of system.jobs -- namely permission checking.
But that makes me a bit sad (cause I'm not a fan of crdb_internal.jobs)

Long time ago, I tried rewriting show jobs, and I was able to emulate permission checking directly in sql by doing something like:
```
+       const decodedJobs = `
+SELECT
+   id, status, created,
+   crdb_internal.pb_to_json('cockroach.sql.jobs.jobspb.Payload', payload) as payload,
+   crdb_internal.pb_to_json('cockroach.sql.jobs.jobspb.Progress', progress) as progress
+FROM system.jobs

+  permissions AS (
+    SELECT
+      crdb_internal.is_admin() as isAdmin,
+      crdb_internal.has_role_option('CONTROLJOB') as hasJobControl
+  ),
+  jobs AS (
+    SELECT *, %s
+    FROM (WITH decodedJobs AS (%s %s) SELECT * FROM decodedJobs), permissions
+    WHERE permissions.isAdmin OR permissions.hasJobControl OR payload->>'username' = current_user()
+  )

```

I don't know.  Let me think a bit more about this query, and whether we should just not use crdb_internal.jobs.
For now, let's address other comments, but we might have to come back to this particular query.
</blockquote></details>

If we're reconsidering using crdb_internal.jobs, should we also consider it's use in SHOW JOBS? (maybe in another ticket)


<!-- Sent from Reviewable.io -->

spiffyy99 reviewed (commented):
> Dismissed @miretskiy from a discussion.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64956)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @spiffyyeng, and @stevendanna)

___
*[pkg/ccl/changefeedccl/show_changefeed_jobs_test.go, line 204 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/64956#-Ma9TOYI3kl_nRtH5g53:-Ma9hb-ZAnspMs2Jo4dJ:b-jn5s6n) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ce2439c0444adeff194f1bf8dacbbb8e6fe428f/pkg/ccl/changefeedccl/show_changefeed_jobs_test.go#L204)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

you probably want to run `show changefeed` statements again after pausing/cancelling (and also you need to resume too).
You should verify that show changefeed job does something reasonable.  Otherwise, why bother pausing/resuming?  You're not 
testing jobs system here.
</blockquote></details>

Discussed offline.


<!-- Sent from Reviewable.io -->

spiffyy99 reviewed (commented):
> Dismissed @miretskiy from 5 discussions.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64956)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @spiffyyeng, and @stevendanna)

___
*[pkg/sql/delegate/show_changefeed_jobs.go, line 38 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/64956#-Ma9B4VG8--adWOJfEoH:-MaA4kNp17dnfZLGkub3:bss4a1x) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ce2439c0444adeff194f1bf8dacbbb8e6fe428f/pkg/sql/delegate/show_changefeed_jobs.go#L38)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

probably not needed since you know it's a changefeed job?
</blockquote></details>

Removed 
___
*[pkg/sql/sem/tree/stmt.go, line 1254 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/64956#-Ma9AFi58cgwZ_aSH9_z:-MaA4uChCLMaEgae6639:bdiogf1) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ce2439c0444adeff194f1bf8dacbbb8e6fe428f/pkg/sql/sem/tree/stmt.go#L1254)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

`SHOW CHANGEFEED JOBS`?
</blockquote></details>

Nice catch


<!-- Sent from Reviewable.io -->

miretskiy reviewed (changes requested):
> Reviewed 3 of 3 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64956)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @spiffyyeng, and @stevendanna)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/64956#-M_QG3WE8GYcolIvI8-K:-MaTatev6QmFhLJYkg-H:br0bpat) (no related file):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Thanks for working on this! I think we and our customers are going to get a lot of use out of it going forward.

I left a few minor comments that I hope aren't too annoying.  Overall though it looks pretty reasonable.  I didn't review the actual query being used too closely since we are going to iterate on what information gets included there over time.
</blockquote></details>

I think it's fine to use crdb_internal.jobs for now.  Just small minor comments.
Let's see if @stevendanna  has more to say on this PR.

---
___
*[pkg/sql/delegate/show_changefeed_jobs.go, line 24 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/64956#-MaTa1x5CuC97OrJpDdb:-MaTa1x5CuC97OrJpDdc:b-nhup2l) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5d9e219afea5105ae2146d3475f2cbeaf1a3152/pkg/sql/delegate/show_changefeed_jobs.go#L24)):*
> ```Go
> 	sqltelemetry.IncrementShowCounter(sqltelemetry.Jobs)
> 
> 	// sink_uri has sensitive information redacted when JSON is marshaled from PB
> ```

Do you mind expanding this comment?
Something like:
`Note: changefeed_details contains sensitive information in sink_url.  This information is redacted by implementing MarshalJSONPB.`

Also, perhaps we should move that method here?


<!-- Sent from Reviewable.io -->

spiffyyeng commented:
@stevendanna any more comments?

miretskiy reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> -- Feel free to update your test as needed.

stevendanna reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64956)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @miretskiy, @spiffyyeng, and @stevendanna)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/64956#-MaZ14_l7JH8_JnlPmey:-MaZ14_l7JH8_JnlPmez:bg7igkv) (no related file):*
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---
___
*[pkg/ccl/changefeedccl/show_changefeed_jobs_test.go, line 114 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/64956#-MaYzQXBCdiyhxvzGGY8:-MaYzQXCB7m_mr_WT2Hl:blc7lwq) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b3c5e4533dee8fb191f772e6ea6211b9fe2051b/pkg/ccl/changefeedccl/show_changefeed_jobs_test.go#L114)):*
> ```Go
> 	var changefeedID jobspb.JobID
> 
> 	// Cannot use kafka for tests right now because of leaked goroutine issue
> ```

Do we have an open issue for this?
___
*[pkg/ccl/changefeedccl/show_changefeed_jobs_test.go, line 125 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/64956#-MaZ-hmt63Y8fW1Nq4Fc:-MaZ-hmt63Y8fW1Nq4Fd:b-1r4u53) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b3c5e4533dee8fb191f772e6ea6211b9fe2051b/pkg/ccl/changefeedccl/show_changefeed_jobs_test.go#L125)):*
> ```Go
> 
> 	require.Equal(t, changefeedID, out.id, "Expected id:%d but found id:%d", changefeedID, out.id)
> 	require.Equal(t, "experimental-s3://fake-bucket-name/fake/path?AWS_ACCESS_KEY_ID=123&AWS_SECRET_ACCESS_KEY=redacted", out.SinkURI, "Expected sinkUri:%s but found sinkUri:%s", "experimental-s3://fake-bucket-name/fake/path?AWS_ACCESS_KEY_ID=123&AWS_SECRET_ACCESS_KEY=redacted", out.SinkURI)
> ```

I think that require.Equal will already produce pretty nice output that contains both the expected and actual values, so you can probably avoid doing that manually if you'd like.
___
*[pkg/sql/delegate/show_changefeed_jobs.go, line 79 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/64956#-MaZ0jIX-DtPWLHl0-PE:-MaZ0jIX-DtPWLHl0-PF:b-xkt1v9) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b3c5e4533dee8fb191f772e6ea6211b9fe2051b/pkg/sql/delegate/show_changefeed_jobs.go#L79)):*
> ```Go
> 		// - then all completed jobs sorted in order of completion time.
> 		whereClause = fmt.Sprintf(
> 			`WHERE %s AND (finished IS NULL OR finished > now() - '12h':::interval)`, typePredicate)
> ```

We should be sure to document this 12 hour bit.  I like the idea of limiting it by default.


<!-- Sent from Reviewable.io -->

spiffyy99 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64956)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @miretskiy, @spiffyyeng, and @stevendanna)

___
*[pkg/ccl/changefeedccl/show_changefeed_jobs_test.go, line 114 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/64956#-MaYzQXBCdiyhxvzGGY8:-MaZ9vOQ2Rg9cMWJSANs:bfrgdpe) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b3c5e4533dee8fb191f772e6ea6211b9fe2051b/pkg/ccl/changefeedccl/show_changefeed_jobs_test.go#L114)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Do we have an open issue for this?
</blockquote></details>

no, but i can create one
___
*[pkg/sql/delegate/show_changefeed_jobs.go, line 79 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/64956#-MaZ0jIX-DtPWLHl0-PE:-MaZ9nVF-0hJ624KkcwR:b-pyiojn) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b3c5e4533dee8fb191f772e6ea6211b9fe2051b/pkg/sql/delegate/show_changefeed_jobs.go#L79)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

We should be sure to document this 12 hour bit.  I like the idea of limiting it by default.
</blockquote></details>

it's on our website for `SHOW JOBS`: https://www.cockroachlabs.com/docs/v20.2/show-jobs.html#considerations 


<!-- Sent from Reviewable.io -->

stevendanna commented:
> it's on our website for SHOW JOBS

Ah, cool, thanks!

spiffyy99 committed:
sql: Add support for SHOW CHANGEFEED JOBS

Previously, SHOW JOBS did not contain enough changefeed information
for users.
SHOW CHANGEFEED JOBS only shows changefeed jobs as well as the full
table name the changefeed monitors, the changefeed sink URI and the
format as computed columns.

Resolves #64933

Release note (sql change): Added new SHOW CHANGEFEED JOBS command with
additional information about changefeeds for improved user visibility.

spiffyyeng commented:
bors r+
